[readobj][Arm][AArch64] Refactor Build Attributes parsing under ELFAtributeParser and add support for AArch64 Build Attributes

sivan-shani opened the pull request:
Refactor readobj to integrate AArch64 Build Attributes under ELFAttributeParser. ELFAttributeParser now serves as a base class for:
- ELFCompactAttrParser, handling Arm-style attributes with a single build attribute subsection.
- ELFExtendedAttrParser, handling AArch64-style attributes with multiple build attribute subsections. This improves code organization and better aligns with the attribute parsing model.

Add support for parsing AArch64 Build Attributes.

sivan-shani committed:
[readobj][Arm][AArch64] Refactor build attribute parsing under ELFAttributeParser and add support for AArch64 Build Attributes

Refactor readobj to integrate AArch64 build attributes under ELFAttributeParser.
ELFAttributeParser now serves as a base class for:
- ELFCompactAttrParser, handling Arm-style attributes with a single build attribute subsection.
- ELFExtendedAttrParser, handling AArch64-style attributes with multiple build attribute subsections.
This improves code organization and better aligns with the attribute parsing model.

Add support for parsing AArch64 Build Attributes.

sivan-shani committed:
Remove '-o -' from llvm-mc arguments since it is the default.

ostannard commented on llvm/include/llvm/Support/ELFAttrParserExtended.h:
```diff
@@ -0,0 +1,51 @@
+  SmallVector<BuildAttributeSubSection, 64> SubSectionVec;
+  // Maps SubsectionName + Tag to tags names. Required for printing comments.
+  const std::vector<SubsectionAndTagToTagName> TagsNamesMap;
+  StringRef getTagName(const StringRef &s, const unsigned i);
```
These argument names should be more descriptive.

ostannard commented on llvm/lib/Support/AArch64AttributeParser.cpp:
```diff
@@ -0,0 +1,19 @@
+
+#include "llvm/Support/AArch64AttributeParser.h"
+
+const std::vector<llvm::SubsectionAndTagToTagName>
```
Could this be a static array, line the ARM version of this, to avoid the extra copying of the `std::vector`?

sivan-shani replied:
It isn't too bad - a single copy when the object is created, and it happens one per run.

Still, possible improvement: The function does not have to return 'const', it does not make any functional change (the important part is `ELFExtendedAttrParser` holding `TagsNamesMap` as a const) therefore, removing the const qualifier should allow the compiler to make a RVO optimization when required (e.g. using move)

Making this change, let me know what you think.

ostannard commented on llvm/lib/Support/ELFAttrParserCompact.cpp:
```diff
@@ -128,9 +131,8 @@ Error ELFAttributeParser::parseSubsection(uint32_t length) {
   }
 
   // Handle a subsection with an unrecognized vendor-name by skipping
-  // over it to the next subsection. ADDENDA32 in the Arm ABI defines
```
Accidental deletion?

sivan-shani replied:
Intentional, since this is not Arm specific.

ostannard replied:
It's still the spec which originally defined this format, so it's useful information for anyone reading this code.

sivan-shani committed:
Provide default implementations for ELFAttributeParser virtual functions
to prevent compilation errors when no derived class is initialized,
such as on non-ELF platforms.

ostannard commented on llvm/lib/Support/ELFAttrParserExtended.cpp:
```diff
@@ -0,0 +1,176 @@
+           (OffsetInSection + ExtBASubsectionLength - BytesAllButAttributes)) {
+
+      uint64_t Tag = De.getULEB128(Cursor);
+      std::string Str = utostr(Tag);
```
This is confusing due to the three similar variables (`Str`, `TagStr`, `TagAsString`), maybe it could be simplified to something like this:
```
std::string TagName;
if (StringRef KnownTagName = getTagName(VendorName, Tag))
  TagName = KnownTagName;
else
  TagName = utostr(Tag);
```

ostannard commented on llvm/lib/Support/ELFAttrParserExtended.cpp:
```diff
@@ -0,0 +1,176 @@
+      if ("" != TagAsString)
+        TagStr = TagAsString;
+
+      uint64_t ValueInt = 0;
```
These aren't modified, and we always construct the `BuildAttributeItem` below with the default values.

ostannard reviewed (commented):
This is another PR with multiple logical changes in a single large patch, it would have been easier to review if split up appropriately. I think there are three changes here:
* Rename `AArch64BuildAttrs` to `AArch64BuildAttributes` (NFC)
* Split out the existing code into `ELFAttrParserCompact` (NFC)
* Add `ELFAttrParserExtended`, and use it for AArch64

sivan-shani committed:
Delete tests_ra.sh

sivan-shani committed:
Delete tests.sh

ostannard commented on llvm/include/llvm/Support/ELFAttrParserExtended.h:
```diff
@@ -0,0 +1,51 @@
+  DataExtractor::Cursor Cursor{0};
+
+  // Data structure for holding Extended ELF Build Attribute subsection
+  SmallVector<BuildAttributeSubSection, 64> SubSectionVec;
```
That test case is a fairly extreme case, because it's trying to test lots of assembler/readobj behaviour, not representative of real usage. 8 isn't an unreasonable value though, so that would be fine too.

sivan-shani committed:
Overload getAttributeValue to accept unsigned only

ELFCompactAttrParser: Add a check for empty string in getAttributeValue

- Overloaded getAttributeValue to accept only an unsigned parameter.
- Added a check in ELFCompactAttrParser to ensure the string is empty when using the getAttributeValue version that accepts both a string and an unsigned.

sivan-shani committed:
Reduce SubSectionVec capacity from 64 to 8 subsections

- Modified SubSectionVec to hold a maximum of 8 subsections instead of 64.

sivan-shani committed:
Rename (const StringRef &s, const unsigned i) --> (const StringRef &BuildAttrSubsectionName, const unsigned Tag) (NFC)
Rename entry -> Entry (NFC)

sivan-shani committed:
Delete tests.sh

sivan-shani committed:
Delete tests_ra.sh

sivan-shani committed:
Remove const qualifier from returnTagsNamesMap()

- No functional effect as ELFExtendedAttrParser holds TagsNamesMap as const.
- Enables RVO optimization (e.g., move instead of copy).

sivan-shani committed:
- Add new line at end of file (NFC)
- Restore comment with mention of ADDENDA32

sivan-shani committed:
Improve unclear and unnecessarily long code

sivan-shani committed:
Assign values that are left unassigned

sivan-shani committed:
test: specify linux to ensure elf in order to solve windows failure

sivan-shani committed:
upstream test failure: give input file explicitly to llvm-mc

sivan-shani committed:
Fix usage of out-of-scope variable.
Change TagsNamesMap to be static.

sivan-shani committed:
Add implementaion of overloaded versions of getAttributeValue, getAttributeString in ELFAttrParserExtended that do not take StringRef
with assertion in case they are being used by mistake.

ostannard reviewed (approved):
LGTM

sivan-shani commented:
Merge cause some failing sanitizers.
fix is here:
https://github.com/llvm/llvm-project/pull/130576
